test/library/standard/DataFrames/DataFrames.chpl

```diff
@@ -539,15 +616,116 @@ module DataFrames {
       return || reduce this.these();
     }
 
+    proc nrows(): int {
+      if idx then
+        return idx.nrows();
+      else
+        return ords.size;
+    }
+
     proc writeThis(f) {
       if idx {
         idx.writeThis(f, this);
       } else {
         for (i, d) in this.items() do
-          f <~> i + "\t" + d + "\n";
+          f <~> i + "    " + d + "\n";
       }
       f <~> "dtype: " + eltType:string;
     }
+
+    pragma "no doc"
+    proc writeElem(f, i, len: int) {
+      var output = if this.valid(i)
+                   then new string(this[i]: string)
+                   else "None";
+
+      for space in 1..len-output.length do
+        f <~> " ";
+      f <~> output;
+    }
+
+    pragma "no doc"
+    proc writeElemNoIndex(f, i: int, len: int) {
+      var output = if this.valid_at(i)
+                   then new string(this.at(i): string)
+                   else "None";
+
+      for space in 1..len-output.length do
+        f <~> " ";
+      f <~> output;
+    }
+  }
+
+  class DataFrame {
+    var labels: domain(string);
+    var columns: [labels] Series;
+    var idx: Index;
+
+    // TODO: init with labels arg
+
+    proc init(columns: [?D] Series) {
+      this.labels = D;
+      this.columns = columns;
+      this.idx = nil;
+    }
+
+    proc init(columns: [?D], idx: Index) {
+      this.labels = D;
+      this.columns = columns;
+      this.idx = idx;
+
+      this.complete();
+      for s in this.columns do
+        s.reindex(idx);
+    }
+
+    iter these() {
+      for s in columns do
+        yield s;
+    }
+
+    proc this(lab: string) {
+      return columns[lab];
+    }
+
+    proc nrows() {
+      var nMax = 0;
+      for s in this {
+        var n = s.nrows();
+        if n > nMax then nMax = n;
+      }
+      return nMax;
+    }
+
+    proc writeThis(f) {
+      if idx {
+        idx.writeThis(f, this);
+      } else {
```

Comment:
I feel like we should create an actual Index if the user doesn't provide one, then we can remove this branch.

Reply:
Right now the practice in the ADTs is that a `nil` idx implies a dense range. #9658 should address this, at which point I agree that we should remove these weird `if idx` branches.